### ISSUE TITLE
Timezone support

### ISSUE THREAD

**Issue by nirenjan** [pbogre/jetlog#32]
A flight departing San Francisco (SFO) on May 1st at 1315 would land in New Delhi (DEL) at 1730 on May 2nd. While this is somewhat related to #31, the duration calculation also needs to take the time zones into account.

The default duration calculation simply takes the difference of the times provided, giving 255 minutes (4h15m), but in reality, the flight would have departed on May 1st at 2015 UTC, and landed on May 2nd at 1200 UTC, giving a total flight time of 15h45m.

**Comment by pbogre** [comment]
I previously open an issue about this. However, implementing it this way (automatically detecting time zones) would mean that people who put departure and arrival date based on the same time zone (which is the way I do it) would get incorrect duration calculation.

I think the best way to solve this would be adding a `arrival_date`, since it fixes this problem for both types of people. Would that work for you?

**Comment by pbogre** [comment]
After looking into this a little bit, the change isn't exactly how I thought. First, the standard for timezone names is "Continent/City", not "City/Country", but that's not the problem. Parsing timezone as a string to `tzinfo` requires at least `pytz`, but since airports don't come with continent data (more on this soon) my initial approach was to try and use their `latitude` and `longitude` values, which means installing another module (`tzwhere`). However I had no success with this module.

But, the airports database from which I created the `airports.db` file in this repo (linked in the README), also has the timezone name as a field, but I removed it because I didn't deem it necessary. I can create a new `airports.db` file that contains that field, implement it in the various models for both backend and frontend, and finally use it when calculating the time difference. I'll look into this soon.

**Comment by nirenjan** [comment]
I did find [timezonefinder](https://pypi.org/project/timezonefinder/) package which will get the timezone for a given latitude/longitude combination, and it's fairly fast. This might be worth considering.

**Comment by nirenjan** [comment]
We don't necessarily need to keep the package as a dependency, but it could be useful enough to get the timezones for the airports that don't have the timezone in the parent DB.

**Comment by pbogre** [comment]
That's a good idea.

If we do implement this feature, how could we still allow people to not use separate time zones for departure/arrival times? Because that's the way I would like to keep doing it.

Maybe a checkmark in the UI that, when enabled, informs the endpoint that it should consider separate time zones. 

Since this issue seems to have increased in complexity, for now I'll keep assuming users use the same time zone for both departure and arrival, and if it gathers interest i'll implement it. I would like to keep this project as simple as it can possibly be, I hope you understand.

**Comment by simonpt** [comment]
@pbogre, while I appreciate that you like to enter departure/arrival times using the departure timezone, it's actually very common to have the departure/arrival times expressed in the local timezones. For example, I just entered some historical flights from old hardcopy itineraries provided by travel agents, and I had to convert the arrival times back to the departure timezone before entering them. Then I realised that daylight saving rules have changed over the years, so I used timeanddate.com to calculate the correct timezone differences for the actual dates of the flights. It actually gets quite complicated, doesn't it? Got me wondering how you would code for all that.

For some really old flights, all I had was rubber stamp marks in old passports (no departure/arrival times), so I really appreciated being able skip the times in Jetlog and just enter a duration, which I calculated by googling "flight time from a to b". Hey, perhaps a future version of Jetlog can estimate the flight time...

I like your idea of a checkmark in the UI to change the interpretation of the arrival time's timezone. But, yeah, good luck working out the correct time difference.

May I say how impressed I am with what you've created. It very nicely helps satisfy my desire to log my historical flights and produce some cool stats. I tried the same with a highly-rated mobile app, but turned out that it only lets you log flights that it knows about, so if you go too far back in time, you can't enter the flight. No such problems with Jetlog. Thanks!

**Comment by jamess60** [comment]
I think ultimately, it should be all flight times in local time (as it would be in the real world), Jetlog is timezone aware, but an option to override start/end time and add a manual duration instead could be useful. 

Regarding the Continent/City issue, there is a complete list of global airports and their timezones [here](https://github.com/lxndrblz/Airports/blob/main/airports.csv)

**Comment by pbogre** [comment]
One way to (at least temporarily) solve this problem might be to add a duration input when creating a flight. This input would show the automatic calculation value, and if it is not correct (i.e., you want it to actually be based on time zones), the user can always set it manually.

Of course this isn't really a fix to the problem, but it does remove all the complications of automatic time zone detection. The main issue I have with this feature is that supporting both modes of input (using the same time zone for both times vs using the time zones corresponding to the airport) becomes quite messy. 

All of this is to get a different method of duration calculation, because ultimately the arrival/departure times don't really matter. For these reasons I think I'll implement the duration input I described above as a temporary fix to this problem. If anyone has something to add to this, please do :)

**Comment by cfarrell987** [comment]
@pbogre I was going to open an issue regarding TZ detection, I think it's an interesting issue; Some users will do what you do; input using the same TZ or in my case use the TZ for departure and arrival which leads to messy durations. Allowing the user to set a flag in settings where they could choose whether they use 1 TZ, dest/arrival TZ or even UTC(Zulu time). and by converting the TZ to Zulu (accounting for +/- 1 day) could solve the issue for the most part. Just a though but would like to see this implemented in some capacity. I'm more than happy to experiment with solutions.

**Comment by pbogre** [comment]
Hello @cfarrell987 thanks for chiming in. I have attempted to account for different timezones in the backend before, and between python packages and implementations it is actually a lot messier than I thought to support various timezones, the hardest part comes from detecting what timezone an airport is in given its coordinates.

Ultimately, the only difference that different time zones makes is in the calculation of the flight's duration right? Because the user will know whether they use the same time zones for arrival and departure or if they only use the departure TZ, or whatever other system they want to use. That's why I think simply providing an input for custom duration is the most elegant and simple solution for this issue.

However if you want to contribute and manage to implement TZ detection, I'd be happy to rethink about this.

**Comment by denverquane** [comment]
Just stumbled onto this project; thanks for your hard work!

I almost always use local times (localized to the arrival/dept airport) for flight recording, in part because that's what tends to display in emails and flight confirmations.

I've used [this dataset on Github](https://github.com/lxndrblz/Airports/blob/main/airports.csv) to great success for the timezone issue; any thoughts on this? Guess I'm also not too sure of the ramifications of including the dataset in entirety into the project and/or if it would live in the backend or frontend. 

I personally like the idea of a simple checkbox next to the time inputs that denotes if the times should be treated as "Local Times", but I understand @pbogre you might have different thoughts on that vs. the simple duration input field. I just got very tired of calculating/finding the duration myself 😅

**Comment by pbogre** [comment]
Thanks for the comment @denverquane!

I think this issue is definitely one that i'm going to address next since it's so requested, and I understand that many people use localized times to log their flights.

That dataset you linked looks similar to the one I am using. Could you describe how you were able to find the timezone a certain airport is in using that dataset? I have tried using coordinates but to no avail. Perhaps one way could be to use the city, but I would need some package or database that stores that information. 

Despite the timezone detection problem, and i'm leaving this here more as a note to myself than anything, it could be useful to display the TZ next to departure/arrival times e.g. when viewing a flight (I hope that's doable UI-wise without too much clutter); but then also have an option to hide the TZ for people like me who don't care about them.

**Comment by denverquane** [comment]
> Could you describe how you were able to find the timezone a certain airport is in using that dataset?

Oh I just used the IATA code to lookup the airport, and then parsed the timezone string associated with it. The data is really nice, for my purposes, I was literally just calculating the flight duration so I didn't need anything else besides a IATA ->timezone lookup. It could be dumped to a DB easily, or just kept as a CSV, either would work.

**Comment by cfarrell987** [comment]
Just a thought, but it would it be better to store the timezone in the database alongside the airport? Would take a lot of upfront computation to do so but would greatly reduce it for everyone in the long run.

**Comment by pbogre** [comment]
@denverquane Thanks for the clarifications. What worries me is that using that dataset specifically would mean having another airport db migration which could be a problem especially if that dataset has less airports, which it seems to have.

@cfarrell987 That's definitely one option, and although it would make the database update more complicated it also makes sense and is the fastest performing. The other option is to retrieve an airport's timezone when a flight's arrival/departure times are set/edited, which is slower but keeps the same db as we have now.

Before I even think about which of this options is better though, I would need to find a way to actually compute the airports' time zones, perhaps using their `municipality` or `region` fields is a good way to do that through some python package , but i'd have to look into it. In the meantime i'm open to suggestions.

**Comment by denverquane** [comment]
> Before I even think about which of this options is better though, I would need to find a way to actually compute the airports' time zones

Did you look at the date csv? It literally has the timezones for every airport, not sure I understand the obstacle there.

**Comment by pbogre** [comment]
Yeah I looked at the CSV and my only worry is that many airports in Jetlog's database are not there, even though it has more entries than Jetlog's. I just ran a query and it shows that there are 3179 airports in Jetlog's db that are not in that one, so I either ignore all of those and leave them without timezone or migrate to that db altogether. I'm reluctant to migrate because of the possible issues that could cause with more than 3000 missing airports. So either I merge them and leave the 3000+ airports with a null timezone, or figure out a way to compute the timezones myself.

**Comment by jamess60** [comment]
"leave the 3000+ airports with a null timezone, or figure out a way to compute the timezones myself."

Surely this is as simple as a quick python script with pandas and airport-codes (or one of the many other similar modules) to read the CSV, if timezone = null, populate it, export csv

**Comment by pbogre** [comment]
Yeah my problem was with figuring out how to populate it, but since you mention an `airport-codes` module that's exactly what I've been looking for. I can't find the one called `airport-codes`, but I have found [`airportsdata`](https://pypi.org/project/airportsdata/), which seems to have what I need. I'm gonna try using this module and see how much coverage I can get with this + the database you provided, and I'll see where I can go from there. Thanks for the suggestion :)

**Comment by pbogre** [comment]
Good news! After experimenting a little with python packages, I found [`timezonefinder`](https://pypi.org/project/timezonefinder/), which allows conversion from coordinates to timezone in string format! I managed to update the airports database using this simple python script:

```py
import sqlite3
from timezonefinder import TimezoneFinder

tf = TimezoneFinder()

# open db
con = sqlite3.connect("./data/airports.db")
cur = con.cursor()
cur.execute("SELECT icao, latitude, longitude FROM airports;")
res = cur.fetchall()

# for each airport
for r in res:
    icao = r[0]
    lat = r[1]
    lon = r[2]

    # get tz
    tz = tf.timezone_at(lat=lat, lng=lon)
    assert type(tz) == str

    # set tz in database
    cur.execute("UPDATE airports SET timezone = ? WHERE icao=?;", [tz, icao])
    cur.fetchone()
    con.commit()
```

Currently I have an almost-working branch with this feature. Since each airport now has an associated timezone, all that's left to do is change the method used to calculate duration to consider the airport's timezone. I have opened a draft PR in the meantime.

**Comment by pbogre** [comment]
I'm happy to announce that this feature has finally been implemented. I will make a release soon, but for now you should be able to test it using the `:experimental` image

**Comment by jamess60** [comment]
Excellent work, thanks!
